gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -626,20 +625,46 @@ public void rename(URI src, URI dst) throws IOException {
       throw new FileNotFoundException("Item not found: " + src);
     }
 
+    Optional<CoopLockOperationRename> coopLockOp =
+        options.isCooperativeLockingEnabled()
+                && src.getAuthority().equals(dst.getAuthority())
+                && srcInfo.isDirectory()
+            ? Optional.of(CoopLockOperationRename.create(gcs, pathCodec, src, dst))
+            : Optional.empty();
+    coopLockOp.ifPresent(CoopLockOperationRename::lock);
+    if (coopLockOp.isPresent()) {
+      fileInfo = getFileInfos(paths);
+      srcInfo = fileInfo.get(0);
+      dstInfo = fileInfo.get(1);
+      if (!srcInfo.exists()) {
+        coopLockOp.ifPresent(CoopLockOperationRename::unlock);
+        throw new FileNotFoundException("Item not found: " + src);
+      }
+      if (!srcInfo.isDirectory()) {
+        coopLockOp.ifPresent(CoopLockOperationRename::unlock);
+        coopLockOp = Optional.empty();
+      }
+    }
+
+    FileInfo dstParentInfo = dstParent == null ? null : fileInfo.get(2);
+
     // Throw if src is a file and dst == GCS_ROOT
     if (!srcInfo.isDirectory() && dst.equals(GCS_ROOT)) {
+      coopLockOp.ifPresent(CoopLockOperationRename::unlock);
```

Comment:
Avoid code duplication like this with try/finally? 
Alternatively, you can implement AutoLockable with AutoCloseable interface.

Reply:
Done.